From ef54536aa27b2f5af27db73f1728d75e1770c0a5 Mon Sep 17 00:00:00 2001
From: Limeng <Meng.Li@windriver.com>
Date: Mon, 21 Nov 2016 10:55:28 +0800
Subject: [PATCH 3/3] drivers: mmc: implement wifi+bt feature

on customer's demo board, there is a wifi+bt module(AW-NB177NF)
connected with SD1 and uart3 interface, do below modification to
support the wifi+bt module.
1. Enable wifi node in dts file.
2. Add gpio property in dts file to power on and enable wifi+bt module.
3. Enable uart3 node in dts file, because SoC communicate with bt module
   via uart3.
4. Set SD1 and uart3 dividing parameter in clock initialization module.
5. Enable SD1 clock when set SD1 interface in probe() function.
6. Open compile switch of wifi driver.

Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../boot/dts/actduino_bubble_gum_sdboot_linux.dts  |   17 +++++-------
 arch/arm/mach-owl/clocktree-atm7059.c              |    6 ++++
 drivers/mmc/host/gl520x_mmc.c                      |    5 +++
 drivers/net/wireless/actions/rtl8723bs/Makefile    |    2 +-
 include/linux/wlan_plat.h                          |   27 ++++++++++++++++++++
 5 files changed, 46 insertions(+), 11 deletions(-)
 create mode 100644 include/linux/wlan_plat.h

diff --git a/arch/arm/boot/dts/actduino_bubble_gum_sdboot_linux.dts b/arch/arm/boot/dts/actduino_bubble_gum_sdboot_linux.dts
index 1757d79..d288063 100755
--- a/arch/arm/boot/dts/actduino_bubble_gum_sdboot_linux.dts
+++ b/arch/arm/boot/dts/actduino_bubble_gum_sdboot_linux.dts
@@ -452,7 +452,7 @@
 		pinctrl-names = "default";
 		pinctrl-0 = <&serial3_state_default>;
 //		actions,enable-dma-rx;
-		status = "disabled";
+		status = "okay";
 	};
 
 	serial@b0128000 {
@@ -986,15 +986,9 @@
 		pinctrl-names = "default";
 		pinctrl-0 = <&mmc1_state_default>;
 
-		card_type = "memory";  /* only: "memory", "wifi", or "emmc" */
-		card_detect_mode = "command";  /* only: "sirq", "gpio", or "command" */
-//		card_detect_gpios = <&gpio 125 0>; /*hr280ac GPIOD29 */  
-		interrupts = < 0 43 0x4 >;
-
-		status = "disabled";
+		card_type = "wifi";  /* only: "memory", "wifi", or "emmc" */
 
-		sd_vcc = "switch1";
-//		sdio_uart_supported;
+		status = "okay";
 	};
 
 	mmc@b0238000 {
@@ -1162,7 +1156,10 @@
 	
 	wifi_bt_power_ctl {
 		compatible = "wifi,bt,power,ctl";
-		no_bind_usb_port;
+		wifi_bt_power_gpios = <&gpio 116 0>; /* GPIOD20 */
+		bt_en_gpios = <&gpio 32 0>;/*GPIOB0*/
+		wifi_en_gpios = <&gpio 34 0>; /* GPIOB2 */
+		bt_power_gpio = <&gpio 113 0>; /* GPIOD17 */
 	};
 
 	gsensor_detect {
diff --git a/arch/arm/mach-owl/clocktree-atm7059.c b/arch/arm/mach-owl/clocktree-atm7059.c
index 9d6e29d..6e9018a 100755
--- a/arch/arm/mach-owl/clocktree-atm7059.c
+++ b/arch/arm/mach-owl/clocktree-atm7059.c
@@ -2729,6 +2729,12 @@ void atm7059_init_clocktree(struct device_node *cum_node)
 		if (clocks[i].actdiv) {
 			clocks[i].divsel = read_clkreg_val(clocks[i].actdiv->reg);
 			clocks[i].divider = getdivider(clocks[i].actdiv, clocks[i].divsel);
+			if (i == CLOCK__PRESD1_CLK)
+				clocks[i].divider = clocks[CLOCK__PRESD0_CLK].divider;
+
+			if (i == CLOCK__UART3_CLK)
+				clocks[i].divider = clocks[CLOCK__UART2_CLK].divider;
+
 			if (clocks[i].divider < 0) {
 				clocks[i].divsel = getdivider_resetval(clocks[i].actdiv);
 				write_clkreg_val(clocks[i].actdiv->reg, clocks[i].divsel);
diff --git a/drivers/mmc/host/gl520x_mmc.c b/drivers/mmc/host/gl520x_mmc.c
index ceb8597..b24d4b6 100755
--- a/drivers/mmc/host/gl520x_mmc.c
+++ b/drivers/mmc/host/gl520x_mmc.c
@@ -1865,6 +1865,11 @@ static int __init acts_mmc_probe(struct platform_device *pdev)
 			pr_err("please choose card detect method\n");
 		}
 	} else if (mmc_card_expected_wifi(host->type_expected)) {
+		struct clk *sd1_clk;
+		sd1_clk = clk_get(NULL, "CMUMOD_SD1");
+		clk_prepare(sd1_clk);
+		clk_enable(sd1_clk);
+
 		mmc->caps &= ~MMC_CAP_NEEDS_POLL;
 		res = platform_get_resource(pdev, IORESOURCE_IRQ, 0);
 		if (!res) {
diff --git a/drivers/net/wireless/actions/rtl8723bs/Makefile b/drivers/net/wireless/actions/rtl8723bs/Makefile
index 336b419..3e44095 100755
--- a/drivers/net/wireless/actions/rtl8723bs/Makefile
+++ b/drivers/net/wireless/actions/rtl8723bs/Makefile
@@ -783,7 +783,7 @@ endif
 
 ifeq ($(CONFIG_PLATFORM_ACTIONS_ATM705X), y)
 EXTRA_CFLAGS += -DCONFIG_LITTLE_ENDIAN
-#EXTRA_CFLAGS += -DRTW_ENABLE_WIFI_CONTROL_FUNC
+EXTRA_CFLAGS += -DRTW_ENABLE_WIFI_CONTROL_FUNC
 # default setting for Android 4.1, 4.2, 4.3, 4.4
 EXTRA_CFLAGS += -DCONFIG_PLATFORM_ACTIONS_ATM705X
 ifndef CONFIG_PLATFORM_UBUNTU
diff --git a/include/linux/wlan_plat.h b/include/linux/wlan_plat.h
new file mode 100644
index 0000000..40ec348
--- /dev/null
+++ b/include/linux/wlan_plat.h
@@ -0,0 +1,27 @@
+/* include/linux/wlan_plat.h
+ *
+ * Copyright (C) 2010 Google, Inc.
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ */
+#ifndef _LINUX_WLAN_PLAT_H_
+#define _LINUX_WLAN_PLAT_H_
+
+struct wifi_platform_data {
+	int (*set_power)(int val);
+	int (*set_reset)(int val);
+	int (*set_carddetect)(int val);
+	void *(*mem_prealloc)(int section, unsigned long size);
+	int (*get_mac_addr)(unsigned char *buf);
+	void *(*get_country_code)(char *ccode);
+};
+
+#endif
-- 
1.7.5.4

